tachyon/crypto/sumcheck/sumcheck_verifier.h

```diff
@@ -0,0 +1,204 @@
+    randomness_.push_back(random_value);
+    polynomials_received_.push_back(prover_msg.evaluations);
+
+    if (round_ == num_variables_) {
+      finished_ = true;
+    } else {
+      ++round_;
+    }
+    return {std::move(random_value)};
+  }
+
+  // Verifies the sumcheck phases' validity and generates a subclaim.
+  //
+  // If the asserted sum is correct, then the multilinear polynomial evaluated
+  // at |subclaim.point| is |subclaim.expected_evaluation|. Otherwise, it is
+  // highly unlikely that these two will be equal. A larger field size
+  // guarantees a smaller soundness error.
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // |CheckAndGenerateSubclaim()| verifies for each round i:
+  //   gᵢ(0) + gᵢ(1) =  gᵢ₋₁(rᵢ₋₁)
+  // Note: For the first round, gᵢ₋₁(rᵢ₋₁) = H = |LinearCombination.Combine()|
+  //   Meanwhile, for the last round, gᵢ₋₁(rᵢ₋₁) = the |LinearCombination|
+  //   evaluated on the multivariate point of Verifier's challenges.
+  bool CheckAndGenerateSubclaim(const F& asserted_sum, Subclaim<F>& result) {
+    // Verifier has not finished.
+    CHECK(finished_);
+    // Insufficient rounds.
+    CHECK_EQ(polynomials_received_.size(), num_variables_);
+
+    F expected = asserted_sum;
+    for (size_t i = 0; i < num_variables_; ++i) {
+      const std::vector<F>& evaluations = polynomials_received_[i];
+      // Incorrect number of evaluations.
+      CHECK_EQ(evaluations.size(), max_evaluations_ + 1);
+
+      const F p0 = evaluations[0];
+      const F p1 = evaluations[1];
+
+      F sum = p0 + p1;
+      // Check that prover message is consistent with the claim.
+      if (sum != expected) {
+        return false;
+      }
+
+      expected = InterpolateUniPoly(evaluations, randomness_[i]);
+    }
+    result = {std::move(randomness_), std::move(expected)};
+    return true;
+  }
+
+ private:
+  size_t num_variables_ = 0;
+  size_t max_evaluations_ = 0;
+  size_t round_ = 1;
+  bool finished_ = false;
+  // A list storing the randomness sampled by the verifier at each round.
+  std::vector<F> randomness_;
+  // A list storing the univariate polynomial in evaluation form sent by the
+  // prover at each round.
+  std::vector<std::vector<F>> polynomials_received_;
+};
+
+template <typename F>
+F FieldFactorial(const size_t a) {
+  F res = F::One();
+  F offset = F::One();
+  for (size_t i = 2; i <= a; ++i) {
+    offset += F::One();
+    res *= offset;
+  }
+  return res;
+}
+
+// Interpolates the unique univariate polynomial |poly| of degree at most
+// |poly.size()| - 1 = |poly_size| passing through the y-values in |poly| at
+// x = 0,..., |poly[i].size()| - 1 and evaluates this polynomial at
+// |evaluation_point|:
+//   ∑ poly[i] * ∏ⱼ≠ᵢ(|evaluation_point| - j)/(i - j), (i = 0,1,...,|poly_size|)
+template <typename F>
+F InterpolateUniPoly(const std::vector<F>& poly, const F& evaluation_point) {
+  size_t poly_size = poly.size();
+  // Assumming |poly_size| is type of "size_t".
```

Comment:
```suggestion
  // Assuming |poly_size| is type of "size_t".
```

Reply:
Plus, I don't understand the meaning of the comment.

Reply:
While I was thinking deeply about the related logic, I realized I wasn't sure if poly_size should be fully bound by a size_t -> it stands of max_evals so it may be made bigger? Currently it's constrained as a size_t, though, so I'll just remove that comment.